Attestz-3

dipchauh opened the pull request:
"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

dipchauh committed:
Attestz-3

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

dplore reviewed (commented):
Please move to /feature/security/gnsi folder. 

@mihirpitale-googler to assign reviewers

dipchauh committed:
Merge branch 'main' of github.com:openconfig/featureprofiles into attestzdot3

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

dipchauh committed:
Update nokia pcr verify

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

dipchauh committed:
Merge branch 'main' of github.com:openconfig/featureprofiles into attestzdot3

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

betuls commented on feature/security/attestz/tests/attestz3/README.md:
```diff
@@ -0,0 +1,54 @@
+
+## Summary
+
+In TPM enrollment workflow the switch owner verifies device's Initial Attestation Key (IAK) and Initial DevID (IDevID) certificates (signed by the switch vendor CA) and installs/rotates owner IAK (oIAK) and owner IDevID (oIDevID) certificates (signed by switch owner CA). In TPM attestation workflow switch owner verifies that the device's end-to-end boot state (bootloader, OS, secure boot policy,  ...
```
nit:Also list initial instruction and BIOS code in the examples.

betuls commented on feature/security/attestz/tests/attestz3/README.md:
```diff
@@ -0,0 +1,54 @@
+| attestz-3.1 | Successful post-install re-attestation relying an owner-issued mTLS cert | Device passed attestation for all control cards relying on the latest oIAK and mTLS certs |
+| attestz-3.2 | Two re-attestations separated by a device reboot result in the same PCR values, but different PCR Quote (due a different random nonce in `AttestRequest`) | Device passed multiple re-attestations separated by a reboot for all control cards relying on the latest oIAK and mTLS certs |
+| attestz-3.3 | When an active control card becomes unavailable, standby control card becomes active and can successfully complete re-attestation | Standby control card passed re-attestation after an active control card failure, relying on the latest oIAK and mTLS certs|
+| attestz-3.4 | Device is unable to authenticate switch owner (e.g. no suitable TLS trust bundle) during attestation | `Attest` returns authentication failure error |
```
Are these the initial set of tests? Are we planning on addressing the missing tests later? For example there should be tests that test attestaion fails if secure boot is disabled or the kernel singing key is not in the key database.

betuls commented on feature/security/attestz/tests/attestz3/README.md:
````diff
@@ -0,0 +1,54 @@
+```yaml
+rpcs:
+   gnsi:
+      certz.v1.Certz.Rotate:
````
This needs update, attestz is missing.

dipchauh replied:
Readme parsing fixed in https://github.com/openconfig/featureprofiles/pull/5133

betuls commented on internal/security/attestz/attestz.go:
```diff
@@ -0,0 +1,531 @@
+	t.Logf("Found software version: %v", ver)
+
+	// Expected pcr values for Nokia present in /mnt/nokiaos/<build binary.bin>/known_good_pcr_values.json.
+	sshC, err := dut.RawAPIs().BindingDUT().DialCLI(context.Background())
```
Whats the point of verifying PCRs from the same device? One from gRPC one from direct read from device.
I think even hardcoding the values in a test json file and verifying againt them would be more robust. If there is a vendor system to connect and read the values from it'd be excelent.

dipchauh committed:
Merge branch 'main' of github.com:openconfig/featureprofiles into attestzdot3

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

dipchauh committed:
update package name

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

dipchauh committed:
add support for ecdsa signatures

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

dipchauh committed:
update readme

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

dipchauh committed:
Merge branch 'main' of github.com:openconfig/featureprofiles into attestzdot3

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

dipchauh committed:
vendor ca handling for nokia

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."

trathod1 committed:
Merge branch 'main' into attestzdot3

dipchauh committed:
Merge branch 'main' into attestzdot3

dipchauh committed:
fix test registry

"This code is a Contribution to the OpenConfig Feature Profiles project ("Work") made under the Google Software Grant and Corporate Contributor License Agreement ("CLA") and governed by the Apache License 2.0. No other rights or licenses in or to any of Nokia's intellectual property are granted for any other purpose. This code is provided on an "as is" basis without any warranties of any kind."
